### ISSUE TITLE
Streaming with custome filter Android

### ISSUE THREAD

**Issue by tuanpdPaditech** [pedroSG94/RootEncoder#1612]
I want to integrate deepAR with face effect. i try use custom GLSurfaceView.Renderer
this is my code:
```
package com.example.streaming_deepar.streaming

import ai.deepar.ar.DeepAR
import android.app.Activity
import android.graphics.SurfaceTexture
import android.opengl.EGL14
import android.opengl.GLES11Ext
import android.opengl.GLES20
import android.opengl.GLSurfaceView
import android.opengl.GLSurfaceView.EGLContextFactory
import android.opengl.Matrix
import android.util.Log
import android.view.Surface
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer
import java.nio.ShortBuffer
import javax.microedition.khronos.egl.EGL10
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.egl.EGLContext
import javax.microedition.khronos.egl.EGLDisplay
import javax.microedition.khronos.opengles.GL10


class DeepARRenderer(
    private val deepAR: DeepAR,
    private val rtmpStream: RtmpDeepAR,
    private val context: Activity
) : GLSurfaceView.Renderer {
    private val vertexShaderCode = "attribute vec4 vPosition;" +
            "attribute vec2 vUv;" +
            "varying vec2 uv; " +
            "void main() {" +
            "gl_Position = vPosition;" +
            "uv = vUv;" +
            "}"

    private val fragmentShaderCode = """
        #extension GL_OES_EGL_image_external : require
        precision mediump float;varying vec2 uv; uniform samplerExternalOES sampler;void main() {  gl_FragColor = texture2D(sampler, uv); }
        """.trimIndent()

    private val drawOrder = shortArrayOf(0, 1, 2, 0, 2, 3)
    private val TAG = "DeepARRenderer"

    private var vertexBuffer: FloatBuffer? = null
    private var uvbuffer: FloatBuffer? = null
    private var drawListBuffer: ShortBuffer? = null
    private var program = 0
    private var texture = 0
    private var surfaceTexture: SurfaceTexture? = null
    private var surface: Surface? = null

    private var updateTexImage = false

    var isCallInProgress: Boolean = false

    private var mEGLCurrentContext: EGLContext? = null

    private var textureWidth = 0
    private var textureHeight = 0
    private lateinit var byteBuffer: ByteBuffer

    var matrix: FloatArray = FloatArray(16)
    private lateinit var h264Encoder: H264Encoder
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)

    override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
        GLES20.glClearColor(1.0f, 0.0f, 0.0f, 1.0f)

        val bb = ByteBuffer.allocateDirect(squareCoords.size * 4)
        bb.order(ByteOrder.nativeOrder())
        vertexBuffer = bb.asFloatBuffer()
        vertexBuffer?.put(squareCoords)
        vertexBuffer?.position(0)

        val bb2 = ByteBuffer.allocateDirect(squareCoords.size * 4)
        bb2.order(ByteOrder.nativeOrder())
        uvbuffer = bb2.asFloatBuffer()
        uvbuffer?.put(uv)
        uvbuffer?.position(0)

        val dlb = ByteBuffer.allocateDirect(drawOrder.size * 2)
        dlb.order(ByteOrder.nativeOrder())
        drawListBuffer = dlb.asShortBuffer()
        drawListBuffer?.put(drawOrder)
        drawListBuffer?.position(0)

        val vertexShader = loadShader(GLES20.GL_VERTEX_SHADER, vertexShaderCode)
        val fragmentShader = loadShader(GLES20.GL_FRAGMENT_SHADER, fragmentShaderCode)

        program = GLES20.glCreateProgram()
        GLES20.glAttachShader(program, vertexShader)
        GLES20.glAttachShader(program, fragmentShader)
        GLES20.glLinkProgram(program)

        Matrix.setIdentityM(this.matrix, 0)
    }

    override fun onSurfaceChanged(gl: GL10, width: Int, height: Int) {
        GLES20.glViewport(0, 0, width, height)

        val textures = IntArray(1)
        GLES20.glGenTextures(1, textures, 0)
        texture = textures[0]

        GLES20.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, texture)

        textureWidth = width
        textureHeight = height

        GLES20.glTexImage2D(
            GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
            0,
            GLES20.GL_RGBA,
            textureWidth,
            textureHeight,
            0,
            GLES20.GL_RGBA,
            GLES20.GL_UNSIGNED_BYTE,
            null
        )
        GLES20.glTexParameteri(
            GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
            GLES20.GL_TEXTURE_MIN_FILTER,
            GLES20.GL_LINEAR
        )
        GLES20.glTexParameteri(
            GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
            GLES20.GL_TEXTURE_MAG_FILTER,
            GLES20.GL_LINEAR
        )
        GLES20.glTexParameteri(
            GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
            GLES20.GL_TEXTURE_WRAP_S,
            GLES20.GL_CLAMP_TO_EDGE
        )
        GLES20.glTexParameteri(
            GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
            GLES20.GL_TEXTURE_WRAP_T,
            GLES20.GL_CLAMP_TO_EDGE
        )

        surfaceTexture = SurfaceTexture(texture)
        surface = Surface(surfaceTexture)
        h264Encoder = H264Encoder(textureWidth, textureHeight)
        Log.d(TAG, "width: $textureWidth, height: $textureHeight")
        surfaceTexture!!.setOnFrameAvailableListener { updateTexImage = true }
        // 4 bytes per pixel (RGBA)
        byteBuffer = ByteBuffer.allocateDirect(textureWidth * textureHeight * 4).apply {
                order(ByteOrder.nativeOrder())
            }

        context.runOnUiThread { deepAR.setRenderSurface(surface, textureWidth, textureHeight) }
    }

    override fun onDrawFrame(gl: GL10) {
        GLES20.glFinish()
        GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT or GLES20.GL_DEPTH_BUFFER_BIT)
        if (updateTexImage) {
            updateTexImage = false
            synchronized(this) {
                surfaceTexture!!.updateTexImage()
            }
        }

        surfaceTexture!!.getTransformMatrix(matrix)

        GLES20.glUseProgram(program)
        val positionHandle = GLES20.glGetAttribLocation(program, "vPosition")
        GLES20.glEnableVertexAttribArray(positionHandle)
        GLES20.glVertexAttribPointer(positionHandle, 3, GLES20.GL_FLOAT, false, 12, vertexBuffer)

        val uvHandle = GLES20.glGetAttribLocation(program, "vUv")
        GLES20.glEnableVertexAttribArray(uvHandle)
        GLES20.glVertexAttribPointer(uvHandle, 2, GLES20.GL_FLOAT, false, 8, uvbuffer)

        val sampler = GLES20.glGetUniformLocation(program, "sampler")
        GLES20.glUniform1i(sampler, 0)

        GLES20.glActiveTexture(GLES20.GL_TEXTURE0)
        GLES20.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, texture)

        GLES20.glDrawElements(
            GLES20.GL_TRIANGLES,
            drawOrder.size,
            GLES20.GL_UNSIGNED_SHORT,
            drawListBuffer
        )

        GLES20.glDisableVertexAttribArray(positionHandle)
        GLES20.glDisableVertexAttribArray(uvHandle)
        GLES20.glUseProgram(0)
        if (isCallInProgress) {
            scope.launch {
                try {
                   // Đảm bảo ByteBuffer được đặt lại trước khi đọc
                    // Đọc pixel từ framebuffer
                    GLES20.glReadPixels(
                        0, 0, textureWidth, textureHeight,
                        GLES20.GL_RGBA, GLES20.GL_UNSIGNED_BYTE, byteBuffer
                    )
                    byteBuffer.rewind()
                    val data = h264Encoder.encodeRgbaToH264(byteBuffer)
                    val h264 = data?.first
                    val info = data?.second
                    Log.d(TAG, "h264: $h264, info: $info")
                    if (info != null && h264 != null) {
                        rtmpStream.sendVideo(h264, info)
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    class MyContextFactory(private val renderer: DeepARRenderer) : EGLContextFactory {
        override fun createContext(egl: EGL10, display: EGLDisplay, config: EGLConfig): EGLContext {
            val attrib_list = intArrayOf(
                EGL14.EGL_CONTEXT_CLIENT_VERSION, 2,
                EGL10.EGL_NONE
            )
            renderer.mEGLCurrentContext =
                egl.eglCreateContext(display, config, EGL10.EGL_NO_CONTEXT, attrib_list)

            return renderer.mEGLCurrentContext!!
        }

        override fun destroyContext(
            egl: EGL10, display: EGLDisplay,
            context: EGLContext
        ) {
            if (!egl.eglDestroyContext(display, context)) {
            }
        }
    }

    companion object {
        var squareCoords: FloatArray = floatArrayOf(
            -1.0f, 1.0f, 0.0f,
            -1.0f, -1.0f, 0.0f,
            1.0f, -1.0f, 0.0f,
            1.0f, 1.0f, 0.0f
        )

        var uv: FloatArray = floatArrayOf(
            0.0f, 0.0f,
            0.0f, 1.0f,
            1.0f, 1.0f,
            1.0f, 0.0f
        )

        fun loadShader(type: Int, shaderCode: String?): Int {
            val shader = GLES20.glCreateShader(type)
            GLES20.glShaderSource(shader, shaderCode)
            GLES20.glCompileShader(shader)
            return shader
        }
    }
}
```
i try convert bytebuffer RGAB by H264Encoder
```
package com.example.streaming_deepar.streaming

import android.media.MediaCodec
import android.media.MediaCodecInfo
import android.media.MediaFormat
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import java.nio.ByteBuffer

class H264Encoder(private val width: Int, private val height: Int) {
    private var mediaCodec: MediaCodec? = null
    private val bufferInfo = MediaCodec.BufferInfo()

  
    private var yuvBuffer: ByteArray =
        ByteArray((width * height * 3) / 2) //

    init {
        setupMediaCodec()
    }


    private fun setupMediaCodec() {
        mediaCodec = MediaCodec.createEncoderByType(MediaFormat.MIMETYPE_VIDEO_AVC).apply {
            val format =
                MediaFormat.createVideoFormat(MediaFormat.MIMETYPE_VIDEO_AVC, width * 2, height * 2)
                    .apply {
                        setInteger(
                            MediaFormat.KEY_COLOR_FORMAT,
                            MediaCodecInfo.CodecCapabilities.COLOR_FormatYUV420SemiPlanar
                        )
                        setInteger(MediaFormat.KEY_BIT_RATE, 125000)  // Thay đổi bitrate nếu cần
                        setInteger(MediaFormat.KEY_FRAME_RATE, 15)    // Giảm FPS xuống
                        setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, 1)
                    }

            configure(format, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
            mediaCodec?.createInputSurface();
            start()
        }
    }

  
    fun encodeRgbaToH264(rgbaBuffer: ByteBuffer): Pair<ByteBuffer?, MediaCodec.BufferInfo?>? {
        var encodedOutput: Pair<ByteBuffer, MediaCodec.BufferInfo>? = null
    
        rgbaToYuv420(rgbaBuffer, yuvBuffer, width, height)
        
        val inputBufferIndex = mediaCodec?.dequeueInputBuffer(0) ?: return null
        if (inputBufferIndex >= 0) {
            val inputBuffer = mediaCodec?.getInputBuffer(inputBufferIndex)!!
            inputBuffer.clear()
            
            val inputBufferCapacity = inputBuffer.capacity()
            val yuvBufferSize = yuvBuffer.size

            if (inputBufferCapacity >= yuvBufferSize) {
                inputBuffer.put(yuvBuffer)
                mediaCodec?.queueInputBuffer(
                    inputBufferIndex,
                    0,
                    yuvBufferSize,
                    System.nanoTime() / 1000,
                    0
                )
            } else {
                Log.e(
                    "H264Encoder",
                    "Input buffer is smaller than YUV buffer! Input Buffer Size: $inputBufferCapacity, YUV Buffer Size: $yuvBufferSize"
                )
                return null
            }
        }

  
        val outputBufferIndex = mediaCodec?.dequeueOutputBuffer(bufferInfo, 0) ?: return null
        if (outputBufferIndex >= 0) {
            val outputBuffer = mediaCodec?.getOutputBuffer(outputBufferIndex)!!
            encodedOutput = Pair(outputBuffer, bufferInfo)
            mediaCodec?.releaseOutputBuffer(outputBufferIndex, false)
        } else {
            Log.e("H264Encoder", "No output buffer available!")
        }

        return encodedOutput
    }


    private fun rgbaToYuv420(
        rgbaBuffer: ByteBuffer,
        yuvBuffer: ByteArray,
        width: Int,
        height: Int
    ) {
        var yIndex = 0
        var uIndex = width * height
        var vIndex = uIndex + (width * height / 4)

        for (i in 0 until height) {
            for (j in 0 until width) {
                val rgbaIndex = (i * width + j) * 4
                val r = rgbaBuffer.get(rgbaIndex).toInt() and 0xFF
                val g = rgbaBuffer.get(rgbaIndex + 1).toInt() and 0xFF
                val b = rgbaBuffer.get(rgbaIndex + 2).toInt() and 0xFF

    
                val y = (0.299 * r + 0.587 * g + 0.114 * b).toInt()
                val u = ((b - y) * 0.565 + 128).toInt()
                val v = ((r - y) * 0.713 + 128).toInt()

        
                yuvBuffer[yIndex++] = y.toByte()

            
                if (i % 2 == 0 && j % 2 == 0) {
                    if (uIndex < yuvBuffer.size) {
                        yuvBuffer[uIndex++] = u.toByte()
                    }
                    if (vIndex < yuvBuffer.size) {
                        yuvBuffer[vIndex++] = v.toByte()
                    }
                }
            }
        }
    }


    fun release() {
        mediaCodec?.stop()
        mediaCodec?.release()
        mediaCodec = null
    }
}
```
with above code, when i received an error when try sendvideo via method rtmpstream.sendvideo()

```
java.lang.IllegalStateException: buffer is inaccessible
	at java.nio.DirectByteBuffer.slice(DirectByteBuffer.java:171)
	at java.nio.DirectByteBuffer.slice(DirectByteBuffer.java:41)
	at com.pedro.common.ExtensionsKt.removeInfo(Extensions.kt:68)
	at com.pedro.rtmp.flv.video.packet.H264Packet.createFlvPacket(H264Packet.kt:72)
	at com.pedro.rtmp.rtmp.RtmpSender.sendVideoFrame(RtmpSender.kt:119)
```
could you help me?
Thank you.

**Comment by pedroSG94** [comment]
Hello,

I think that the error is because the library try to modify a ByteBuffer that is read only.

Try to copy the byte buffer of this line:
```kotlin
            val outputBuffer = mediaCodec?.getOutputBuffer(outputBufferIndex)!!
```
To other ByteBuffer:
```kotlin

fun ByteBuffer.clone(): ByteBuffer = ByteBuffer.wrap(toByteArray())

```
Try to do the same for MediaCodec.BufferInfo, to avoid problems create a new instance and copy the values

**Comment by tuanpdPaditech** [comment]
@pedroSG94 
thanks for support.
I have modified the code at as bellow:
```
                     byteBuffer.rewind()
                    GLES20.glReadPixels(
                        0, 0, textureWidth, textureHeight,
                        GLES20.GL_RGBA, GLES20.GL_UNSIGNED_BYTE, byteBuffer
                    )
                    val h264Buffer = ByteBuffer.wrap(byteBuffer.toByteArray())
                    info.apply {
                        size = h264Buffer.remaining()
                        presentationTimeUs = System.nanoTime() / 1000
                        flags = 0
                    }
                    rtmpStream.sendVideo(h264Buffer, info)
```
when code run i received  a logcat: waiting for a valid sps and pps

i am using custom RtmpStream

```
class RtmpDeepAR(context: Context, connectChecker: ConnectChecker, videoSource: VideoSource,
                 audioSource: AudioSource
): StreamBase(context,videoSource,audioSource) {

    val rtmpClient = RtmpClient(connectChecker)
    private val streamClientListener = object: StreamClientListener {
        override fun onRequestKeyframe() {
            requestKeyframe()
        }
    }
    override fun getStreamClient(): RtmpStreamClient = RtmpStreamClient(rtmpClient, streamClientListener)

    constructor(context: Context, connectChecker: ConnectChecker):
            this(context, connectChecker, Camera2Source(context), MicrophoneSource())

    override fun setVideoCodecImp(codec: VideoCodec) {
        rtmpClient.setVideoCodec(codec)
    }

    override fun setAudioCodecImp(codec: AudioCodec) {
        rtmpClient.setAudioCodec(codec)
    }

    override fun onAudioInfoImp(sampleRate: Int, isStereo: Boolean) {
        rtmpClient.setAudioInfo(sampleRate, isStereo)
    }

    override fun startStreamImp(endPoint: String) {
        val resolution = super.getVideoResolution()
        rtmpClient.setVideoResolution(resolution.width, resolution.height)
        rtmpClient.setFps(super.getVideoFps())
        rtmpClient.connect(endPoint)
    }

    override fun stopStreamImp() {
        rtmpClient.disconnect()
    }

    override fun onVideoInfoImp(sps: ByteBuffer, pps: ByteBuffer?, vps: ByteBuffer?) {
        rtmpClient.setVideoInfo(sps, pps, vps)
    }

    override fun getVideoDataImp(videoBuffer: ByteBuffer, info: MediaCodec.BufferInfo) {
        rtmpClient.sendVideo(videoBuffer, info)
    }

    override fun getAudioDataImp(audioBuffer: ByteBuffer, info: MediaCodec.BufferInfo) {
        rtmpClient.sendAudio(audioBuffer, info)
    }

    fun sendVideo(h264Buffer: ByteBuffer, info: MediaCodec.BufferInfo){
        rtmpClient.sendVideo(h264Buffer, info);
    }

}
```

and create rtmp like below:

```
    private val rtmpStream: RtmpDeepAR by lazy {
        RtmpDeepAR(context = this, connectChecker= this, videoSource = NoVideoSource(), audioSource = MicrophoneSource()).apply {
            getGlInterface().autoHandleOrientation = true
            getStreamClient().setBitrateExponentialFactor(0.5f)
        }
    }
```

can you tell me what wrong?

**Comment by pedroSG94** [comment]
Hello,

The problem is that you are not setting videoInfo to the rtmpClient.
If you are using RtmpStream. I recommend you create a new VideoSource instead of do your current implementation.
For example:
```kotlin
@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
class RGBASource: VideoSource() {

  private var encoder: H264Encoder? = null
  
  override fun create(width: Int, height: Int, fps: Int, rotation: Int): Boolean {
    return true
  }

  override fun start(surfaceTexture: SurfaceTexture) {
    encoder = H264Encoder(width, height, Surface(surfaceTexture))
  }

  override fun stop() {
    encoder?.release()
    encoder = null
  }

  override fun release() { }

  override fun isRunning(): Boolean = encoder != null
  
  fun setVideo(rgbaBuffer: ByteBuffer) {
    encoder?.encodeRgbaToH264(rgbaBuffer)
  }
}

@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
class H264Encoder(private val width: Int, private val height: Int, private val surface: Surface) {
  private var mediaCodec: MediaCodec? = null
  private val bufferInfo = MediaCodec.BufferInfo()


  private var yuvBuffer: ByteArray =
    ByteArray((width * height * 3) / 2) //

  init {
    setupMediaCodec()
  }


  private fun setupMediaCodec() {
    mediaCodec = MediaCodec.createEncoderByType(MediaFormat.MIMETYPE_VIDEO_AVC).apply {
      val format =
        MediaFormat.createVideoFormat(MediaFormat.MIMETYPE_VIDEO_AVC, width * 2, height * 2)
          .apply {
            setInteger(
              MediaFormat.KEY_COLOR_FORMAT,
              MediaCodecInfo.CodecCapabilities.COLOR_FormatYUV420SemiPlanar
            )
            setInteger(MediaFormat.KEY_BIT_RATE, 125000)  // Thay đổi bitrate nếu cần
            setInteger(MediaFormat.KEY_FRAME_RATE, 15)    // Giảm FPS xuống
            setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, 1)
          }
      //render encoded frames in the surface
      configure(format, surface, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
      mediaCodec?.createInputSurface();
      start()
    }
  }


  fun encodeRgbaToH264(rgbaBuffer: ByteBuffer) {
    var encodedOutput: Pair<ByteBuffer, MediaCodec.BufferInfo>? = null

    rgbaToYuv420(rgbaBuffer, yuvBuffer, width, height)

    val inputBufferIndex = mediaCodec?.dequeueInputBuffer(0) ?: return
    if (inputBufferIndex >= 0) {
      val inputBuffer = mediaCodec?.getInputBuffer(inputBufferIndex)!!
      inputBuffer.clear()

      val inputBufferCapacity = inputBuffer.capacity()
      val yuvBufferSize = yuvBuffer.size

      if (inputBufferCapacity >= yuvBufferSize) {
        inputBuffer.put(yuvBuffer)
        mediaCodec?.queueInputBuffer(
          inputBufferIndex,
          0,
          yuvBufferSize,
          System.nanoTime() / 1000,
          0
        )
      } else {
        Log.e(
          "H264Encoder",
          "Input buffer is smaller than YUV buffer! Input Buffer Size: $inputBufferCapacity, YUV Buffer Size: $yuvBufferSize"
        )
        return
      }
    }


    val outputBufferIndex = mediaCodec?.dequeueOutputBuffer(bufferInfo, 0) ?: return
    if (outputBufferIndex >= 0) {
      val outputBuffer = mediaCodec?.getOutputBuffer(outputBufferIndex)!!
      encodedOutput = Pair(outputBuffer, bufferInfo)
      //indicate that we want render the surface
      mediaCodec?.releaseOutputBuffer(outputBufferIndex, true)
    } else {
      Log.e("H264Encoder", "No output buffer available!")
    }
  //we don't need return the output because surface will be handle the output
//    return encodedOutput
  }


  private fun rgbaToYuv420(
    rgbaBuffer: ByteBuffer,
    yuvBuffer: ByteArray,
    width: Int,
    height: Int
  ) {
    var yIndex = 0
    var uIndex = width * height
    var vIndex = uIndex + (width * height / 4)

    for (i in 0 until height) {
      for (j in 0 until width) {
        val rgbaIndex = (i * width + j) * 4
        val r = rgbaBuffer.get(rgbaIndex).toInt() and 0xFF
        val g = rgbaBuffer.get(rgbaIndex + 1).toInt() and 0xFF
        val b = rgbaBuffer.get(rgbaIndex + 2).toInt() and 0xFF


        val y = (0.299 * r + 0.587 * g + 0.114 * b).toInt()
        val u = ((b - y) * 0.565 + 128).toInt()
        val v = ((r - y) * 0.713 + 128).toInt()


        yuvBuffer[yIndex++] = y.toByte()


        if (i % 2 == 0 && j % 2 == 0) {
          if (uIndex < yuvBuffer.size) {
            yuvBuffer[uIndex++] = u.toByte()
          }
          if (vIndex < yuvBuffer.size) {
            yuvBuffer[vIndex++] = v.toByte()
          }
        }
      }
    }
  }


  fun release() {
    mediaCodec?.stop()
    mediaCodec?.release()
    mediaCodec = null
  }
}

```

Now, set this video source to RtmpStream and instead of set frames to RtmpStream in DeepARRenderer, try to send frames to RGBASource

**Comment by tuanpdPaditech** [comment]
@pedroSG94 hi
I have tried many ways but it seems the problem is still not solved
i think the problem lies in the conversion bytebuffer to h264
this is my demo: https://github.com/tuanpdPaditech/demo-deepAR-streaming
please hepl me.
thanks

**Comment by tuanpdPaditech** [comment]
@pedroSG94 when i try to send h264 bytebuffers continuously in onDrawFrame function i get following error:
```
FATAL EXCEPTION: DefaultDispatcher-worker-39
Process: ai.deepar.deepar_example, PID: 1915
java.lang.IllegalStateException: buffer is inaccessible
	at java.nio.DirectByteBuffer.slice(DirectByteBuffer.java:171)
	at java.nio.DirectByteBuffer.slice(DirectByteBuffer.java:41)
	at com.pedro.common.ExtensionsKt.removeInfo(Extensions.kt:68)
	at com.pedro.rtmp.flv.video.packet.H264Packet.createFlvPacket(H264Packet.kt:72)
	at com.pedro.rtmp.rtmp.RtmpSender.sendVideoFrame(RtmpSender.kt:119)
	at com.pedro.rtmp.rtmp.RtmpClient.sendVideo(RtmpClient.kt:546)
```
It seems that by the time removeInfo() is called, the bytebuffer has already been freed by call releaseOutputBuffer()

**Comment by pedroSG94** [comment]
You have a PR in your repo that fix the problem.
After check the code I decided create a VideoSource for DeepAR, this source use surfaces instead of send buffers directly which is much better in performance

**Comment by tuanpdPaditech** [comment]
@pedroSG94 thanks for your support. I will check immediately

**Comment by tuanpdPaditech** [comment]
@pedroSG94  yes, it work fine, thank you very much

**Comment by pedroSG94** [comment]
Added to master, this should be available in the next release 2.5.2

**Comment by tuanpdPaditech** [comment]
@pedroSG94 hi , when i try to stream with DeepARImageFormat.RGBA_8888 it gives error, preview error as well as crash when starting live stream, can you update the code to stream with this format in the next version?
